api/http-api.md

```diff
@@ -168,10 +170,471 @@ Current suggested approach, subject to discussion:
 * Will test all of the above
 * Can discuss approach for cloud in separate place
 
+#### Payload Formatting
+
+One of the major drawbacks of payloads is that if we accept them directly in JSON, they look like:
+
+```json
+{
+  "payloads": [{
+    "metadata": {
+      "encoding": "anNvbi9wbGFpbg=="
+    },
+    "data": "IkhlbGxvLCBXb3JsZCEi"
+  }]
+}
+```
+
+This is for the simple JSON message of `"Hello, World"`. With larger JSON objects it gets even more ugly like if we
+wanted: `{ "foo": "bar" }`, it becomes `"data": "eyAiZm9vIjogImJhciIgfQ=="`. This is because by default proto byte
+arrays are serialized/deserialized as base64.
+
+But we may be able to solve this using a "shorthand" approach. Some research has been done yet, here are the notes:
+
+* grpc-gateway uses `jsonpb` and supports custom marshalling choice (so `gogoproto`'s `jsonpb` could be used instead)
+* We can make custom JSON unmarshalling _per proto_ on these older forms of unmarshalling, e.g. see
+  [here](https://pkg.go.dev/github.com/gogo/protobuf/jsonpb#JSONPBUnmarshaler)
+  * But newer `protojson` does not support this, see [this issue](https://github.com/golang/protobuf/issues/1322) and
+    others. So this could harm our ability to upgrade.
+* We can make `go.temporal.io/api/common/v1.Payloads` impl JSON marshal/unmarshal to assume if it is given a JSON
+  collection, it can assume it's for the "payloads" field
+* We can make `go.temporal.io/api/common/v1.Payload` impl JSON marshal/unmarshal to assume if it not an object that can
+  deserialize into payload using strict rules (e.g. exact `metadata` and `data` format), then treat it as a `json/plain`
+  payload for non `null` or null payload for null value.
+  * We could just say "without metadata assume JSON/null" but what if someone has an object with a "metadata" field?
+  * The only way it will clash with the raw form is if someone happens to have a `metadata` and/or `data` field with
+    proper base64
```

Comment:
> I think any distinction between "shorthand" and "non-shorthand" should be explicit and out-of-band, and I don't think we should try to be clever about detecting one or the other. Otherwise we will inevitably run into corner cases like this one.

I defined the cases fairly clearly here. This edge case will likely never happen and therefore should not justify killing the shorthand approach. Do you have a proposal for how someone should provide a "shorthand" JSON payload that is not base64 encoded?

Are you saying instead of something like:

```bash
curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id -d '{
  "workflowType": { "name": "MyWorkflow" },
  "taskQueue": { "name": "my-task-queue" },
  "input": ["Hello, World!"]
}'
```

Something like:

```bash
curl -X POST http://localhost:7243/api/v1/namespaces/my-namespace/workflows/my-workflow-id -d '{
  "workflowType": { "name": "MyWorkflow" },
  "taskQueue": { "name": "my-task-queue" },
  "input": [{
    "_jsonPayload": true,
    "data": "Hello, World!"
  }]
}'
```

I think that's a big leap. I think just because I pointed out an extreme edge case doesn't mean we should assume it will ever be hit. Granted more research is required to even see if I can do this custom decoding at all.

Reply:
Sure, let's find out if we can do custom decoding first, and once there is a more concrete proposal, we can revisit this.

In general, though, I'm still very suspicious of mixing arbitrary properties with properties that have special behaviors together in the same object because, given enough time and enough users, someone who isn't aware eventually WILL hit it and have a bad day. My suspicion goes even deeper when we are talking about both magic keys AND magic values (e.g. "values must be base64-encoded or a different behavior applies")—those tend to be even harder to debug.

Reply:
I am not planning on a another, more concrete proposal.

> In general, though, I'm still very suspicious of mixing

That makes sense, but we have to consider all options. Can you put forth an alternative proposal for how a payload should look?

Reply:
Roughly what I am thinking is: make "shorthand mode" be something that is turned on or off entirely on a per-request basis, either by an HTTP header or a top-level JSON property in the request body. It would therefore not be possible to mix shorthand and non-shorthand payloads.

This does have the downside of making it harder for users to compose an individual request out of different parts, since the fully-formed request would have to be consistent in its use of shorthand or not. I think this is a worthwhile trade-off because a failure due to "mix-and-match" is more obvious—a user can look at the failed request and see that it mixes shorthand and non-shorthand. Whereas if the server tries to guess, it's not immediately obvious what's happening if nothing in the request itself clues in the user—who didn't read the docs, by the way—that "hey, shorthand is a thing that exists".

I have no opinion on:
- Whether to use an HTTP header or a property in the request body. Some options:
  - Baked into Content-Type: `Content-Type: application/json; temporal-payloads=shorthand`
  - Custom HTTP header: `X-Temporal-Payloads: shorthand`
  - JSON property in request body: `{"_payloads": "shorthand", ...}`
- Whether shorthand mode is on or off by default (i.e. what the behavior is if the payload mode is unspecified)
- The name of the parameter used to enable/disable shorthand mode

If you want more details than this, I'll probably need your help to get a deeper understanding of payloads, the context/places they're used, etc. given that all of this is new to me. :)

Reply:
I think opting in to the clearer approach to payloads is a very bad idea and ruins the benefit. I think you are overestimating the potential for clash. I regret saying there is possibility for ambiguity. Can you give the example of the situation where you would see a clash to justify forcing everyone do to this opt-in? I think you're harming everyone to solve a hypothetical that in reality will never occur. If, in some very strange reality, ambiguity did occur I can show how it would be handled.

I think it's very clear to demonstrate the problem you are seeing before trying to solve it here.

Reply:
> I think opting in to the clearer approach to payloads is a very bad idea and ruins the benefit. [...]

We can make it opt-out (i.e. turn on shorthand by default) instead of opt-in. As I said earlier, I have no opinion on whether it's opt-out or opt-in.

> I think it's very clear to demonstrate the problem you are seeing before trying to solve it here.

You yourself gave an example earlier:

>  * The only way it will clash with the raw form is if someone happens to have a `metadata` and/or `data` field with proper base64

Reply:
> We can make it opt-out (i.e. turn on shorthand by default) instead of opt-in. As I said earlier, I have no opinion on whether it's opt-out or opt-in.

This is hard to do because at deserialization time, we don't have contextual data to make this decision. https://pkg.go.dev/github.com/gogo/protobuf/jsonpb#JSONPBUnmarshaler is what we will implement. There is no external "am I in shorthand mode". We have a possible approach to do this in the https://pkg.go.dev/github.com/gogo/protobuf/jsonpb#JSONPBMarshaler direction because I can https://pkg.go.dev/go.temporal.io/api/proxy#VisitPayloads all the payloads and mark them as "in shorthand mode". I am open to options here though. I can probably cheat by making some kind of custom any resolver and checking that type, but it is ugly.

> You yourself gave an example earlier:

I meant an actual curl call that shows an HTTP call with the clash. I think you'll find it's rare. But at this point, I have explained all I can here.

Reply:
> > We can make it opt-out (i.e. turn on shorthand by default) instead of opt-in. As I said earlier, I have no opinion on whether it's opt-out or opt-in.
> 
> This is hard to do because at deserialization time, we don't have contextual data to make this decision. https://pkg.go.dev/github.com/gogo/protobuf/jsonpb#JSONPBUnmarshaler is what we will implement. There is no external "am I in shorthand mode". We have a possible approach to do this in the https://pkg.go.dev/github.com/gogo/protobuf/jsonpb#JSONPBMarshaler direction because I can https://pkg.go.dev/go.temporal.io/api/proxy#VisitPayloads all the payloads and mark them as "in shorthand mode". I am open to options here though. I can probably cheat by making some kind of custom any resolver and checking that type, but it is ugly.

I understand implementation may be hard/ugly, and if that turns out to be the case, we can always revisit whether it's worth the effort. Do you agree this is ideally what we should be doing, though, or do you have any other concerns?

> I meant an actual curl call that shows an HTTP call with the clash. I think you'll find it's rare. But at this point, I have explained all I can here.

100% agree it would be very rare. My worry is that when someone does run into it, it's going to be extremely frustrating and non-obvious and may cost someone a bunch of time and effort to figure out, and that will be a bad user experience.

Reply:
> Do you agree this is ideally what we should be doing, though, or do you have any other concerns?

I think the devil is in the details. Do you have a suggested way to implement `UnmarshalJSONPB`? How in that method do you know whether you're in "shorthand" mode or not?

Reply:
> > Do you agree this is ideally what we should be doing, though, or do you have any other concerns?
> 
> I think the devil is in the details. Do you have a suggested way to implement `UnmarshalJSONPB`?

I'm intentionally not thinking about implementation details until we are aligned on what the goal actually is. If we agree that, in principle, a per-request flag is preferable to auto-detecting the format for each payload, then we can talk about whether/what/how much implementation complexity is an acceptable trade-off.

Reply:
The goals are clearer when you see how they manifest. Having said that, ok I agree, regardless of implementation we'll add a flag. I don't think it'll ever be used, but at this point it's less work than continuing this thread.